Assets/Scripts/Enemies/BossBehaviorController.cs

```diff
@@ -4,108 +4,156 @@
-        if( bossPhases[currentPhase].startBehaviour != null) bossPhases[currentPhase].startBehaviour.enabled = true;
-        foreach(PhaseAction p in bossPhases[currentPhase].backgroundActions) StartCoroutine( startActionBackground(p) );
-        ChangeAction();
+        if (bossPhases[currentPhase].startBehaviour != null)
+        {
+            bossPhases[currentPhase].startBehaviour.enabled = true;
+            while (bossPhases[currentPhase].startBehaviour.actionRunning)
+            {
+                yield return new WaitForEndOfFrame();
+            }
+            bossPhases[currentPhase].startBehaviour.enabled = false;
+        }
+        else ChangeAction();
+        foreach (PhaseAction p in bossPhases[currentPhase].backgroundActions) StartCoroutine(startActionBackground(p));
+
     }
-    public void ChangeAction () {
-        if(currentAction.behavior != null) currentAction.behavior.enabled = false;
-        if(actionQueue.Any()) { currentAction =  actionQueue[0]; actionQueue.RemoveAt(0); }
-        else if(bossPhases[currentPhase].PossibleActions.Count > 1) 
-            currentAction = bossPhases[currentPhase].PossibleActions.Where(x=> !x.Equals(currentAction)).ElementAt( rand.Next(bossPhases[currentPhase].PossibleActions.Count-1) );
+    public void ChangeAction()
+    {
+        foreach(PhaseAction p in actionQueue) print(p.behavior);
+        if (currentAction.behavior != null) currentAction.behavior.enabled = false;
+        if (bossPhases[currentPhase].repeatingBehaviour != null && currentAction.behavior != bossPhases[currentPhase].repeatingBehaviour) currentAction.behavior = bossPhases[currentPhase].repeatingBehaviour;
+        else if (actionQueue.Any()) { currentAction = actionQueue[0]; actionQueue.RemoveAt(0); }
+        else if (bossPhases[currentPhase].PossibleActions.Count > 1)
+            currentAction = bossPhases[currentPhase].PossibleActions.Where(x => !x.Equals(currentAction)).ElementAt(rand.Next(bossPhases[currentPhase].PossibleActions.Count));
         else currentAction = bossPhases[currentPhase].PossibleActions.FirstOrDefault();
-
-        if(currentAction.behavior != null) StartCoroutine(startAction(currentAction, true));
+        if (currentAction.behavior != null) StartCoroutine(startAction(currentAction, true));
         else Debug.Log("BBC Error: Cannot start null behavior!");
     }
-    public IEnumerator startAction (PhaseAction action, bool startNewOnFinish) {
+    public IEnumerator startAction(PhaseAction action, bool startNewOnFinish)
+    {
         action.behavior.enabled = true;
+        action.behavior.actionRunning = true;
         int startedPhase = currentPhase;
-        for(int i = 0; i < action.runFrames; i++) {
-            if(currentPhase != startedPhase) {
-                action.behavior.enabled = false;
-                yield break;
+        if (action.independent || action.behavior == bossPhases[currentPhase].repeatingBehaviour)
```

Comment:
oh is this for actions which have internal timers? 
that should probably be the default anyway

Reply:
yes lol